sw/device/tests/sim_dv/uart_tx_rx_test.c

```diff
@@ -48,7 +55,28 @@ typedef enum uart_direction {
   kUartReceive,
 } uart_direction_t;
 
-// TODO: Make these datasets random.
+// These datasets will be overwritten by DV test.
+
+// There are multiple uart instances in the chip. DV test will randomly select
+// one and overwrite base address and interrupt IDs below.
+static volatile uint32_t uart_base_addr;
+static volatile uint32_t uart_peripheral_id;
+static volatile uint32_t uart_irq_tx_watermartk_id;
+static volatile uint32_t uart_irq_rx_watermartk_id;
+static volatile uint32_t uart_irq_tx_empty_id;
+static volatile uint32_t uart_irq_rx_overflow_id;
+static volatile uint32_t uart_irq_rx_frame_err_id;
+static volatile uint32_t uart_irq_rx_frame_err_id;
+static volatile uint32_t uart_irq_rx_break_err_id;
+static volatile uint32_t uart_irq_rx_timeout_id;
+static volatile uint32_t uart_irq_rx_parity_err_id;
+
+/**
+ * TODO, create an array instead of a int variable, because we need to overwrite
+ * this from DV to randomly verify one of UART. A int variable will be optimized
+ * and there is not fixed address location for it and DV can't overwrite
+ */
+static const uint8_t uart_idx_array[4] = {0x00, 0x00, 0x00, 0x00};
```

Comment:
Alternately you can declare this as `static volatile const`:
```c
static volatile const uint32_t UartIdx = 0x0;
```

The `volatile prevents the compiler from optimizing this variable. Without it, the compiler thinks that since it is a `const`, it will not be modified so there is no need to even have any of the branching logic in `update_uart_base_addr_and_irq_id()`. But, we do intend to modify it, just that the SW is completely oblivious to how we modify it (i.e. via backdoor overwriting that location in flash).

The `const` is needed to put it in the `.rodata` section, otherwise it gets placed in `.data` section in the main SRAM. We cannot backdoor write anything in SRAM at the start of the test because the CRT init code wipes it to 0s.

Reply:
Excellent!! It works!
Updated, thank you very much.